test/functional/plugin/lsp_spec.lua

```diff
@@ -1039,7 +1039,7 @@ describe('LSP', function()
       'LspDiagnosticsVirtualTextHint',
       'LspDiagnosticsVirtualTextInformation',
       'LspDiagnosticsVirtualTextWarning',
-    }, exec_lua([[require'vim.lsp'; return vim.fn.getcompletion('Lsp', 'highlight')]]))
+    }, exec_lua([[require'vim.diagnostic'; return vim.fn.getcompletion('Lsp', 'highlight')]]))
```

Comment:
might be out of scope for this PR, but I am surprised to see that we have a module where require() does global side-effects. That is an anti-pattern. require() should always be safe.

 Callers should explicitly do something like `require('vim.disagnostic').init()`.

Reply:
The `require()` is needed just to setup highlight groups. It is a global side effect, but it's idempotent. Normally this is never needed since simply using `vim.diagnostic` sets up these highlight groups, but since this is a contrived test that doesn't actually use `vim.diagnostic` but only checks for the highlight groups, we have to do things a little differently.

We could expose some method like `_setup_highlight_groups()` but that would *only* be used in this one test. Not sure that is worth it IMO.

Reply:
Alternatively we could just define the highlight groups in `syntax.c` in the `highlight_init_both` table. That might make sense now that this is a top level module.

Reply:
> We could expose some method like `_setup_highlight_groups()` but that would _only_ be used in this one test. Not sure that is worth it IMO.

Just an `init()` is all that's needed. We're already implicitly doing it.



> but it's idempotent.

Yeah but people copy the pattern and then start doing other things. Also, even idempotent effects are surprising. Simpler to have a single guideline: require() should not have side-effects.

Reply:
What do you think of the suggestion of defining the groups directly in `syntax.c`? That would remove the need for any kind of init() altogether.

Reply:
Either is fine, I think -- for LSP, we also did not predefine highlight groups and explicitly mentioned in the documentation that these need to be defined by the user.

However, that _did_ lead to repeated questions since not everybody reads the documentation...

So I would prefer if these were just defined in `syntax.c` (as they are now no longer LSP-specific) so that colorschemes can redefine them easily.

Reply:
> Either is fine, I think -- for LSP, we also did not predefine highlight groups and explicitly mentioned in the documentation that these need to be defined by the user.

This is not correct. I copied this pattern from the existing code:

https://github.com/neovim/neovim/blob/afdc9e61502a9f02c8cf5156652a1bde027e4345/runtime/lua/vim/lsp.lua#L1538

So already, accessing `vim.lsp` has the same side effects that `vim.diagnostic` would have (i.e. defining highlight groups).

Just wanted to clarify that I'm not introducing anything new here, just sticking with the status quo.

Reply:
> This is not correct. I copied this pattern from the existing code:

Ah, that's relatively new. The documentation still claims that you need to define them (which really did use to be the case, at least for some, before that was changed).

In any case, defining them in `syntax.c` is much cleaner IMHO.

Reply:
> What do you think of the suggestion of defining the groups directly in `syntax.c`? That would remove the need for any kind of init() altogether.

No, I don't think we should start mixing parts of the LSP framework into the C core.

Reply:
But this is no longer LSP-specific but a more general diagnostic framework? (The argument still applies, of course, albeit more weakly.)